File: algorithms/math/modular-exponentiation/kotlin/ModExp.kt
fun modExp(arr: IntArray): Int {
    var base = arr[0].toLong()
    var exp = arr[1].toLong()
    val mod = arr[2].toLong()
    if (mod == 1L) return 0
    var result = 1L
    base %= mod
    while (exp > 0) {
        if (exp % 2 == 1L) {
            result = (result * base) % mod
        }
        exp = exp shr 1
        base = (base * base) % mod
    }
    return result.toInt()
}

File: algorithms/math/modular-exponentiation/python/mod_exp.py
def mod_exp(arr: list[int]) -> int:
    base, exp, mod = arr[0], arr[1], arr[2]
    if mod == 1:
        return 0
    result = 1
    base = base % mod
    while exp > 0:
        if exp % 2 == 1:
            result = (result * base) % mod
        exp = exp >> 1
        base = (base * base) % mod
    return result
